sdk/tools/whitelister/whitelister.cpp

```diff
@@ -0,0 +1,301 @@
+// e.g.:
+// E:\amd64\reactos>git blame -L 699,699 --show-number --show-name modules\rostests\apitests\advapi32\LockServiceDatabase.c
+// 8400fdc0786b rostests/apitests/advapi32/LockDatabase.c 701 (Thomas Faber 2012-06-29 11:48:35 +0000 699)     ok_err(ERROR_SUCCESS);
+//
+// To reverse it use
+// reactos>git blame -L <Old Line>,<Old Line> --reverse <old hash> <old file>
+// e.g.
+// E:\amd64\reactos>git blame -L 701,701 --reverse 8400fdc0786b rostests\apitests\advapi32\LockDatabase.c
+// f2bc1f0e1118 modules/rostests/apitests/advapi32/LockServiceDatabase.c (Roman Masanin 2021-07-24 21:23:58 +0300 701)     ok_err(ERROR_SUCCESS);
+//
+
+struct WL_LINE_ENTRY
+{
+    int Line;
+    string Reason;
+};
+
+struct WL_ENTRY
+{
+    string FullFileName;
+    string ShortFileName;
+    vector<WL_LINE_ENTRY> Lines;
+};
+
+string
+Execute(const char* cmd)
+{
+    array<char, 128> buffer;
+    string result;
+
+    std::unique_ptr<FILE, decltype(&_pclose)> pipe(_popen(cmd, "r"), _pclose);
+    if (pipe == nullptr)
+    {
+        throw std::runtime_error("popen() failed!");
+    }
+
+    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr)
+    {
+        result += buffer.data();
+    }
+
+    return result;
+}
+
+string
+ExecuteCmd(const char* Format, ...)
+{
+    array<char, _MAX_PATH> buffer;
+    va_list argptr;
+
+    va_start(argptr, Format);
+    _vsnprintf(buffer.data(), buffer.size(), Format, argptr);
+    va_end(argptr);
+
+    return Execute(buffer.data());
+}
+
+string
+git_blame(string &File, int Line)
+{
+    return ExecuteCmd("git blame -L %u,%u --show-number --show-name %s",
+        Line, Line, File.c_str());
+}
+
+string
+git_rev_blame(string &OldHash, string &OldFile, int OldLine)
+{
+    return ExecuteCmd("git blame -L %u,%u -s --reverse -s --show-number %s %s",
+        OldLine, OldLine, OldHash.c_str(), OldFile.c_str());
+}
+
+int
+AddWhitelistLine(string &WhitelistFileName, string &SourceFileName, int Line)
+{
+    return 0;
+}
+
+void
+AddWhitelistEntry(vector<WL_ENTRY> &Whitelist, string& Line)
+{
+    size_t start, length;
+
+    // Format:
+    // <Hash> <original file name> <original line> "<reason>"
+    // 8400fdc0786b rostests/apitests/advapi32/LockDatabase.c 701 "Fails on WHS"
+
+    // Start with the hash, which is 12 characters
+    string oldHash = Line.substr(0, 12);
+
+    // Get the original file name
+    start = 13;
+    length = Line.substr(start).find(' ');
+    string oldFile = Line.substr(start, length);
+
+    // Get the original line
+    start += length + 1;
+    length = Line.substr(start).find(' ');
+    int oldLine = stoi(Line.substr(start, length));
+
+    // Get the reason
+    start += length + 1;
+    length = Line.length() - start - 2;
+    string reason = Line.substr(start + 1, length);
+
+    // Translate to head
+    string translated = git_rev_blame(oldHash, oldFile, oldLine);
+
+    fprintf(stderr, "Translated: \"%s\"\n", translated.c_str());
```

Comment:
using `cerr` ? (and in all other places where `fprintf(stderr` is used)

Reply:
Why?

Reply:
Be careful what you wish for, or I slap a bunch `` nullptr``  into it.